Sources/SwiftKafka/KafkaConsumer.swift

```diff
@@ -0,0 +1,393 @@
+        offset: Int64 = Int64(RD_KAFKA_OFFSET_END),
+        config: KafkaConfig,
+        logger: Logger
+    ) throws {
+        // Although an assignment is not related to a consumer group,
+        // librdkafka requires us to set a `group.id`.
+        // This is a known issue:
+        // https://github.com/edenhill/librdkafka/issues/3261
+        var config = config
+        try config.set(UUID().uuidString, forKey: "group.id")
+
+        try self.init(
+            config: config,
+            logger: logger
+        )
+        try self.assign(
+            topic: topic,
+            partition: partition,
+            offset: offset
+        )
+    }
+
+    /// Subscribe to the given list of `topics`.
+    /// The partition assignment happens automatically using `KafkaConsumer`'s consumer group.
+    /// - Parameter topics: An array of topic names to subscribe to.
+    private func subscribe(topics: [String]) throws {
+        assert(!closed)
+
+        for topic in topics {
+            rd_kafka_topic_partition_list_add(
+                self.subscribedTopicsPointer,
+                topic,
+                KafkaPartition.unassigned.rawValue
+            )
+        }
+
+        let result = self.client.withKafkaHandlePointer { handle in
+            rd_kafka_subscribe(handle, subscribedTopicsPointer)
+        }
+
+        guard result.rawValue == 0 else {
+            throw KafkaError(rawValue: result.rawValue)
+        }
+    }
+
+    /// Assign the``KafkaConsumer`` to a specific `partition` of a `topic`.
+    /// - Parameter topic: Name of the topic that this ``KafkaConsumer`` will read from.
+    /// - Parameter partition: Partition that this ``KafkaConsumer`` will read from.
+    /// - Parameter offset: The topic offset where reading begins. Defaults to the offset of the last read message.
+    private func assign(
+        topic: String,
+        partition: KafkaPartition,
+        offset: Int64
+    ) throws {
+        assert(!closed)
+
+        guard let partitionPointer = rd_kafka_topic_partition_list_add(
+            self.subscribedTopicsPointer,
+            topic,
+            partition.rawValue
+        ) else {
+            fatalError("rd_kafka_topic_partition_list_add returned invalid pointer")
+        }
+
+        partitionPointer.pointee.offset = offset
+
+        let result = self.client.withKafkaHandlePointer { handle in
+            rd_kafka_assign(handle, self.subscribedTopicsPointer)
+        }
+
+        guard result == RD_KAFKA_RESP_ERR_NO_ERROR else {
+            throw KafkaError(rawValue: result.rawValue)
+        }
+    }
+
+    /// Receive new messages and forward the result to the ``messages`` `AsyncSequence`.
+    func produceMore() {
+        self.serialQueue.async {
+            guard !self.closed else {
+                return
+            }
+
+            let messageResult: Element
+            do {
+                guard let message = try self.poll() else {
+                    self.produceMore()
```

Comment:
@FranzBusch 
Is this behaviour correct or will the `NIOAsyncSequence` invoke `produceMore()` again anyway?

Reply:
I think for the shape that we are doing here it is okay to call `self.produceMore()` again. The `NIOAsyncSequenceProducer` will only call `produceMore` once when an element has been consumed and you need to make sure that at some point that leads to a new message.